cmd/svcat/completion/completion_cmd.go

```diff
@@ -0,0 +1,129 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package completion
+
+import (
+	"fmt"
+	"github.com/kubernetes-incubator/service-catalog/cmd/svcat/command"
+	"github.com/spf13/cobra"
+)
+
+const defaultBoilerPlate = `
+# Copyright 2017 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+`
+
+var (
+	completionLong = `
+Output shell completion code for the specified shell (bash).
+The shell code must be evaluated to provide interactive
+completion of svcat commands. This can be done by sourcing it from
+the .bash_profile.
+
+Note: this requires the bash-completion framework, which is not installed
+by default on Mac. This can be installed by using homebrew:
+
+	$ brew install bash-completion
+
+Once installed, bash_completion must be evaluated. This can be done by adding the
+following line to the .bash_profile
+
+	$ source $(brew --prefix)/etc/bash_completion
+`
+
+	completionExample = `
+# Install bash completion on a Mac using homebrew
+brew install bash-completion
+printf "\n# Bash completion support\nsource $(brew --prefix)/etc/bash_completion\n" >> $HOME/.bash_profile
+source $HOME/.bash_profile
+
+# Load the svcat completion code for bash into the current shell
+source <(svcat completion bash)
+
+# Write bash completion code to a file and source if from .bash_profile
+svcat completion bash > ~/.svcat/svcat_completion.bash.inc
+printf "\n# Svcat shell completion\nsource '$HOME/.svcat/svcat_completion.bash.inc'\n" >> $HOME/.bash_profile
+source $HOME/.bash_profile
+`
+)
+
+var (
+	completionShells = map[string]func(cxt *command.Context, cmd *cobra.Command) error{
+		"bash": runCompletionBash,
+	}
+)
+
+// NewCompletionCmd return command for executing "svcat completion" command
+func NewCompletionCmd(cxt *command.Context, boilerPlate string) *cobra.Command {
+	shells := []string{}
+	for s := range completionShells {
+		shells = append(shells, s)
+	}
+
+	cmd := &cobra.Command{
+		Use:     "completion SHELL",
+		Short:   "Output shell completion code for the specified shell (bash or zsh).",
+		Long:    completionLong,
+		Example: completionExample,
+		Run: func(cmd *cobra.Command, args []string) {
```

Comment:
svcat uses some helper methods to make the commands easier to unit test. The patter is:

1. Wire up RunE and PreRunE using the helper methods:

    https://github.com/kubernetes-incubator/service-catalog/blob/4dfd13c2b48e5fb33ee048d5c332320bd2502790/cmd/svcat/instance/touch_cmd.go#L42-L43

1. Implement a validate function that parses the args and saves them to the command struct or returns an error

    https://github.com/kubernetes-incubator/service-catalog/blob/4dfd13c2b48e5fb33ee048d5c332320bd2502790/cmd/svcat/instance/touch_cmd.go#L50-L57

1. Implement a run function that uses the fields on the command struct to execute the command

    https://github.com/kubernetes-incubator/service-catalog/blob/4dfd13c2b48e5fb33ee048d5c332320bd2502790/cmd/svcat/instance/touch_cmd.go#L59-L62

Since you need the original cobra command in your run function, save it to the command struct here so you can access it later.